// amocrm-js/dist/api/activeRecords/mixins/embedded/hasCatalogElements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasElements = void 0;
function hasElements(Base) {
    return class HasEmbeddedElements extends Base {
        get catalogElementList() {
            if (this._catalogElementList) {
                return this._catalogElementList;
            }
            this._catalogElementList = {
                get: this.getElements.bind(this),
                add: this.addElements.bind(this),
                remove: this.removeElements.bind(this)
            };
            return this._catalogElementList;
        }
        getElements() {
            var _a;
            return ((_a = this._embedded) === null || _a === void 0 ? void 0 : _a.catalog_elements) || [];
        }
        addElements(criteria) {
            const embedded = this._embedded || {};
            const { catalog_elements = [] } = embedded;
            const factory = this.getFactory();
            const entityCriteria = factory.getEntityCriteria(criteria);
            catalog_elements.push(...entityCriteria);
            this._embedded = Object.assign(Object.assign({}, embedded), { catalog_elements });
            return this;
        }
        removeElements(criteria) {
            const embedded = this._embedded || {};
            const embeddedElements = embedded.catalog_elements || [];
            const ids = criteria.map(({ id }) => id);
            const catalog_elements = embeddedElements.filter(({ id }) => !id || !ids.includes(id));
            this._embedded = Object.assign(Object.assign({}, embedded), { catalog_elements });
            return this;
        }
    };
}
exports.hasElements = hasElements;
