// amocrm-js/dist/api/activeRecords/mixins/embedded/hasContacts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasContacts = void 0;
function hasContacts(Base) {
    return class HasEmbeddedContacts extends Base {
        get contactList() {
            if (this._contactList) {
                return this._contactList;
            }
            this._contactList = {
                get: this.getContacts.bind(this),
                add: this.addContacts.bind(this),
                remove: this.removeContacts.bind(this)
            };
            return this._contactList;
        }
        getContacts() {
            var _a;
            return ((_a = this._embedded) === null || _a === void 0 ? void 0 : _a.contacts) || [];
        }
        addContacts(criteria) {
            const embedded = this._embedded || {};
            const { contacts = [] } = embedded;
            const factory = this.getFactory();
            const entityCriteria = factory.getEntityCriteria(criteria);
            contacts.push(...entityCriteria);
            this._embedded = Object.assign(Object.assign({}, embedded), { contacts });
            return this;
        }
        removeContacts(criteria) {
            const embedded = this._embedded || {};
            const embeddedContacts = embedded.contacts || [];
            const ids = criteria.map(({ id }) => id);
            const contacts = embeddedContacts.filter(({ id }) => !id || !ids.includes(id));
            this._embedded = Object.assign(Object.assign({}, embedded), { contacts });
            return this;
        }
    };
}
exports.hasContacts = hasContacts;
