// amocrm-js/dist/api/activeRecords/mixins/embedded/hasTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasTags = void 0;
function hasTags(Base) {
    return class HasEmbeddedTags extends Base {
        get tagList() {
            if (this._tagList) {
                return this._tagList;
            }
            this._tagList = {
                get: this.getTags.bind(this),
                add: this.addTags.bind(this),
                remove: this.removeTags.bind(this)
            };
            return this._tagList;
        }
        getTags() {
            var _a;
            return ((_a = this._embedded) === null || _a === void 0 ? void 0 : _a.tags) || [];
        }
        addTags(criteria) {
            const embedded = this._embedded || {};
            const { tags = [] } = embedded;
            const factory = this.getFactory();
            const entityCriteria = factory.getEntityCriteria(criteria);
            tags.push(...entityCriteria);
            this._embedded = Object.assign(Object.assign({}, embedded), { tags });
            return this;
        }
        removeTags(criteria) {
            const embedded = this._embedded || {};
            const ids = criteria.map(({ id }) => id);
            const tags = this.getTags()
                .filter(({ id }) => !id || !ids.includes(id));
            this._embedded = Object.assign(Object.assign({}, embedded), { tags });
            return this;
        }
    };
}
exports.hasTags = hasTags;
