// amocrm-js/dist/api/activeRecords/mixins/hasTasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasTasks = exports.TaskEntityCriteriaItem = void 0;
const EntityList_1 = require("../common/EntityList");
class TaskEntityCriteriaItem {
    constructor(entity) {
        this.entity = entity;
    }
    fetchCriteria() {
        return {
            filter: {
                entity_id: this.entity.id
            }
        };
    }
    createCriteria() {
        return {
            entity_id: this.entity.id
        };
    }
}
exports.TaskEntityCriteriaItem = TaskEntityCriteriaItem;
function hasTasks(Base) {
    return class HasTasks extends Base {
        constructor(entityFactory) {
            super(entityFactory);
            const factory = Object.create(this.getFactory().tasks);
            const criteriaItem = new TaskEntityCriteriaItem(this);
            factory.criteriaBuilder.add(criteriaItem);
            this.tasks = new EntityList_1.EntityList({
                factory,
            });
            this.Task = function (attributes) {
                return factory.from(attributes);
            };
        }
    };
}
exports.hasTasks = hasTasks;
