// amocrm-js/dist/api/activeRecords/Task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Task = exports.mixins = exports.BaseTask = void 0;
const tslib_1 = require("tslib");
const ResourceEntity_1 = tslib_1.__importDefault(require("../ResourceEntity"));
const hasSave_1 = require("./mixins/hasSave");
const hasFetch_1 = require("./mixins/hasFetch");
const util_1 = require("../../util");
const hasCreate_1 = require("./mixins/hasCreate");
const hasUpdate_1 = require("./mixins/hasUpdate");
const hasLinkedEntity_1 = require("./mixins/hasLinkedEntity");
/**
 * Сделка
 */
class BaseTask extends ResourceEntity_1.default {
    getAttributes() {
        return {
            id: this.id,
            created_by: this.created_by,
            updated_by: this.updated_by,
            created_at: this.created_at,
            updated_at: this.updated_at,
            responsible_user_id: this.responsible_user_id,
            group_id: this.group_id,
            entity_id: this.entity_id,
            entity_type: this.entity_type,
            is_completed: this.is_completed,
            task_type_id: this.task_type_id,
            text: this.text,
            duration: this.duration,
            complete_till: this.complete_till,
            result: this.result,
            account_id: this.account_id
        };
    }
    setAttributes(attributes) {
        this.id = attributes.id;
        this.created_by = attributes.created_by;
        this.updated_by = attributes.updated_by;
        this.created_at = attributes.created_at;
        this.updated_at = attributes.updated_at;
        this.responsible_user_id = attributes.responsible_user_id;
        this.group_id = attributes.group_id;
        this.entity_id = attributes.entity_id;
        this.entity_type = attributes.entity_type;
        this.is_completed = attributes.is_completed;
        this.task_type_id = attributes.task_type_id;
        this.text = attributes.text;
        this.duration = attributes.duration;
        this.complete_till = attributes.complete_till;
        this.result = attributes.result;
        this.account_id = attributes.account_id;
    }
}
exports.BaseTask = BaseTask;
exports.mixins = [
    hasCreate_1.hasCreate,
    hasUpdate_1.hasUpdate,
    hasSave_1.hasSave,
    hasFetch_1.hasFetch,
    hasLinkedEntity_1.hasLinkedEntity
];
exports.Task = (0, util_1.applyMixins)(BaseTask, [
    ...exports.mixins
]);
