// amocrm-js/dist/api/factories/common/FactoryCriteriaBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FactoryCriteriaBuilder = exports.CriteriaBuilderType = void 0;
const tslib_1 = require("tslib");
const deepmerge_1 = tslib_1.__importDefault(require("deepmerge"));
var CriteriaBuilderType;
(function (CriteriaBuilderType) {
    CriteriaBuilderType["GET"] = "fetch";
    CriteriaBuilderType["CREATE"] = "create";
    CriteriaBuilderType["UPDATE"] = "update";
    CriteriaBuilderType["DELETE"] = "delete";
})(CriteriaBuilderType = exports.CriteriaBuilderType || (exports.CriteriaBuilderType = {}));
class FactoryCriteriaBuilder {
    constructor(factory) {
        this.factory = factory;
        this.items = [];
    }
    add(criteriaItem) {
        this.items.push(criteriaItem);
    }
    remove(criteriaItem) {
        this.items = this.items.filter(item => criteriaItem !== item);
    }
    getCriteria(type, defaults) {
        const method = (type + 'Criteria');
        return this.items.reduce((data, factoryCriteriaItem) => {
            const patchHandler = factoryCriteriaItem[method];
            if (!patchHandler) {
                return data;
            }
            const patch = patchHandler.call(factoryCriteriaItem, defaults);
            if (Array.isArray(data)) {
                return data.map(attributes => (0, deepmerge_1.default)(attributes, patch));
            }
            return (0, deepmerge_1.default)(data, patch);
        }, defaults);
    }
}
exports.FactoryCriteriaBuilder = FactoryCriteriaBuilder;
