// amocrm-js/dist/api/factories/mixins/hasTasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasTasks = exports.HasTasksFactoryCriteriaItem = void 0;
const getEntityFactory_1 = require("../common/getEntityFactory");
class HasTasksFactoryCriteriaItem {
    constructor(entityType) {
        this.entityType = entityType;
    }
    fetchCriteria() {
        return {
            entity_type: this.entityType
        };
    }
    createCriteria() {
        return {
            entity_type: this.entityType
        };
    }
    updateCriteria() {
        return {
            entity_type: this.entityType
        };
    }
}
exports.HasTasksFactoryCriteriaItem = HasTasksFactoryCriteriaItem;
function hasTasks(Base) {
    return class HasTasks extends Base {
        constructor(client) {
            super(client);
            const tasks = (0, getEntityFactory_1.getEntityFactory)({
                factory: this,
                type: 'tasks'
            });
            const entityType = this.getEmbeddedKey();
            const criteriaItem = new HasTasksFactoryCriteriaItem(entityType);
            tasks.criteriaBuilder.add(criteriaItem);
            this.tasks = tasks;
        }
    };
}
exports.hasTasks = hasTasks;
