// amocrm-js/dist/api/factories/LeadFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeadFactory = exports.mixins = exports.BaseLeadFactory = void 0;
const tslib_1 = require("tslib");
/**
 * Фабрика для создания сделок {@link Lead}
 * */
const ResourceFactory_1 = tslib_1.__importDefault(require("../ResourceFactory"));
const v4_1 = tslib_1.__importDefault(require("../../schema/v4"));
const hasGetByCriteria_1 = require("./mixins/hasGetByCriteria");
const hasGetById_1 = require("./mixins/hasGetById");
const hasCreate_1 = require("./mixins/hasCreate");
const hasUpdate_1 = require("./mixins/hasUpdate");
const util_1 = require("../../util");
const hasTags_1 = require("./mixins/hasTags");
const hasTasks_1 = require("./mixins/hasTasks");
/**
 * Фабрика управления сделками
 * */
class BaseLeadFactory extends ResourceFactory_1.default {
    getEntityClass() {
        return this.getClient().getEntityConstructors().Lead;
    }
    getBaseUrl() {
        return v4_1.default.entities.leads.path;
    }
    getEmbeddedKey() {
        return 'leads';
    }
    /**
     * @todo https://www.amocrm.ru/developers/content/crm_platform/leads-api#leads-complex-add
     * */
    complexCreate() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return false;
        });
    }
}
exports.BaseLeadFactory = BaseLeadFactory;
exports.mixins = [
    hasGetByCriteria_1.hasGetByCriteria,
    hasGetById_1.hasGetById,
    hasCreate_1.hasCreate,
    hasUpdate_1.hasUpdate,
    hasTags_1.hasTags,
    hasTasks_1.hasTasks
];
exports.LeadFactory = (0, util_1.applyMixins)(BaseLeadFactory, exports.mixins);
