// amocrm-js/dist/api/factories/TaskFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskFactory = exports.mixins = exports.BaseTaskFactory = void 0;
const tslib_1 = require("tslib");
const Task_1 = require("../activeRecords/Task");
const ResourceFactory_1 = tslib_1.__importDefault(require("../ResourceFactory"));
const v4_1 = tslib_1.__importDefault(require("../../schema/v4"));
const util_1 = require("../../util");
const hasGetByCriteria_1 = require("./mixins/hasGetByCriteria");
const hasGetById_1 = require("./mixins/hasGetById");
const hasCreate_1 = require("./mixins/hasCreate");
const hasUpdate_1 = require("./mixins/hasUpdate");
/**
 * Фабрика управления сделками
 * */
class BaseTaskFactory extends ResourceFactory_1.default {
    constructor(client, entityType) {
        super(client);
        this.entityType = entityType;
    }
    getEntityClass() {
        return Task_1.Task;
    }
    getBaseUrl() {
        return v4_1.default.entities.tasks.path;
    }
    getEmbeddedKey() {
        return 'tasks';
    }
}
exports.BaseTaskFactory = BaseTaskFactory;
exports.mixins = [
    hasGetByCriteria_1.hasGetByCriteria,
    hasGetById_1.hasGetById,
    hasCreate_1.hasCreate,
    hasUpdate_1.hasUpdate,
];
exports.TaskFactory = (0, util_1.applyMixins)(BaseTaskFactory, exports.mixins);
