// amocrm-js/dist/api/activeRecords/common/EntityCriteriaBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityCriteriaBuilder = void 0;
const tslib_1 = require("tslib");
const deepmerge_1 = tslib_1.__importDefault(require("deepmerge"));
const util_1 = require("../../../util");
const mergeOptions = {
    arrayMerge: util_1.overwriteMerge
};
class EntityCriteriaBuilder {
    constructor(entity) {
        this.entity = entity;
        this.items = [];
    }
    add(item) {
        this.items.push(item);
    }
    get attributes() {
        return Object.assign(Object.assign({}, this.entity.getAttributes()), { _embedded: {} });
    }
    get createCriteria() {
        return this.items.reduce((attributes, item) => {
            return (0, deepmerge_1.default)(attributes, item.createCriteria, mergeOptions);
        }, this.attributes);
    }
    get updateCriteria() {
        return this.items.reduce((attributes, item) => {
            return (0, deepmerge_1.default)(attributes, item.updateCriteria, mergeOptions);
        }, this.attributes);
    }
}
exports.EntityCriteriaBuilder = EntityCriteriaBuilder;
